slick/src/main/scala/slick/jdbc/DriverDataSource.scala

```diff
@@ -46,42 +46,45 @@ class DriverDataSource(
   // State that gets initialized by `init`
   @volatile private[this] var registered: Boolean = false
   @volatile private[this] var initialized = false
-  private[this] var driver: Driver = _
-  private[this] var connectionProps: Properties = _
+  @volatile private[this] var driver: Driver = _
+  @volatile private[this] var connectionProps: Properties = _
```

Comment:
Hm, right. The `getConnection` methods should be fine because they call `init` first, which reads `initialized` at the beginning and writes it at the end, so that provides the correct ordering. But `close` wouldn't be safe because it only checks `registered`, which is written before `driver` in `init`. Reversing those writes should be sufficient (but I doubt that the performance difference would make any difference).

While checking this I noticed that `getParentLogger` doesn't call `init`. That looks like another bug in the original implementation that needs to be fixed.